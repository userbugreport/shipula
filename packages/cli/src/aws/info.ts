import AWS, { CloudFormation } from "aws-sdk";
import { getStackPath } from "../context";
import path from "path";

/**
 * List all Shipula stacks. This grabs all stacks
 * and filters down by tag.
 */
export const listShipulaStacks = async (): Promise<CloudFormation.Stack[]> => {
  const cloudFormation = new AWS.CloudFormation();
  let buffer = new Array<CloudFormation.Stack>();

  const more = async (
    nextToken: string
  ): Promise<CloudFormation.DescribeStacksOutput> => {
    return cloudFormation
      .describeStacks({ NextToken: nextToken === "-" ? undefined : nextToken })
      .promise();
  };
  let token = "-";
  // more to fetch...
  while (token) {
    const { Stacks, NextToken } = await more(token);
    buffer = [...buffer, ...Stacks];
    token = NextToken;
  }
  // let exceptions leak and be handled by state machines
  // calling this
  return buffer.filter((stack) =>
    stack?.Tags?.find(
      (tag) => tag.Key === "createdBy" && tag.Value === "shipula"
    )
  );
};

/**
 * List all shipula parameters for a given app and stack.
 */
export const listShipulaParameters = async (
  packageName: string,
  stackName: string
): Promise<AWS.SSM.ParameterList> => {
  const ssm = new AWS.SSM();
  let buffer = new Array<AWS.SSM.Parameter>();

  const more = async (
    nextToken: string
  ): Promise<AWS.SSM.GetParametersByPathResult> => {
    return ssm
      .getParametersByPath({
        Path: path.join("/", getStackPath(packageName, stackName)),
        NextToken: nextToken === "-" ? undefined : nextToken,
      })
      .promise();
  };

  let token = "-";
  // more to fetch...
  while (token) {
    const { Parameters, NextToken } = await more(token);
    buffer = [...buffer, ...Parameters];
    token = NextToken;
  }
  return buffer;
};







/**
 * ECS has some batshit names for CPUs. No real person will think that 256 means .25 of a CPU.
 *
 * Spell this out in a big, painful constant.
 *
 * ðŸ§¬ ðŸ§  ?
 */
export const CPU_Memory = [
  {
    display: 1,
    cpu: 1024,
    memory: {
      2: 2048,
      3: 3072,
      4: 4096,
      5: 5120,
      6: 6144,
      7: 7168,
      8: 8192,
    },
  },
  {
    display: 2,
    cpu: 2048,
    memory: {
      4: 4096,
      5: 5120,
      6: 6144,
      7: 7168,
      8: 8192,
      9: 9 * 1024,
      10: 10 * 1024,
      11: 11 * 1024,
      12: 12 * 1024,
      13: 13 * 1024,
      14: 14 * 1024,
      15: 15 * 1024,
      16: 16 * 1024,
    },
  },
  {
    display: 4,
    cpu: 4096,
    memory: {
      8: 8192,
      9: 9 * 1024,
      10: 10 * 1024,
      11: 11 * 1024,
      12: 12 * 1024,
      13: 13 * 1024,
      14: 14 * 1024,
      15: 15 * 1024,
      16: 16 * 1024,
      17: 17 * 1024,
      18: 18 * 1024,
      19: 19 * 1024,
      20: 20 * 1024,
      21: 21 * 1024,
      22: 22 * 1024,
      23: 23 * 1024,
      24: 24 * 1024,
      25: 25 * 1024,
      26: 26 * 1024,
      27: 27 * 1024,
      28: 28 * 1024,
      29: 29 * 1024,
      30: 30 * 1024,
    },
  },
];

/**
 * Decode those goofy CPU values.
 */
export const decodeCPU = (cpu: string): number => {
  return CPU_Memory.find((e) => `${e.cpu}` === cpu).display;
};
