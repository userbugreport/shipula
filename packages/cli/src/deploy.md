Deploy your App to the cloud.

Creates a default Stack, or you can specift a stack name.

This will provision from the network on down to shared disk, with a minimum deployment being:
* A load balancer, with HTTP/HTTPS
* 2 running copies
* 1 CPU
* 8G RAM
* A shared file system at `/var/data`