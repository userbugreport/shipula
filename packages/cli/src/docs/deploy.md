Deploy your App to the cloud.

Creates a default Stack, or you can specify a stack name.

This will provision from the network on down to shared disk, with a minimum deployment being:

- A load balancer, with HTTP/HTTPS
- 2 running copies
- 1 CPU
- 2G RAM
- A shared file system at `/cluster_shared`

You can make as many named stacks as you like -- if you do not specify, it will be called `default`.

You can create a static site, with no running server.
This is automatic if docusaurus is detected, or you can force it by having
a marker in your package.json:

```
"shipula": {
  "static": true
}
```

This will:

- Run your `scripts.build` target in your `package.json`
- Take the contents of `<package_root>/build` and host it as a static site
- `index.html` is the default page, but all pages will be served
