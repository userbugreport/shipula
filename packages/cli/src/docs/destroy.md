This will remove everything from the load balancer through to shared file system, but will leave your backups.
