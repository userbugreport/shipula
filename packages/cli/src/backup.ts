import { Command } from "commander";
import { StackName } from "./constants";
import docs from "./docs";

export default new Command()
  .command("backup [package]")
  .description(
    "Creates a backup copy of your App/Stack file system, and provides a link where you can download it."
  )
  .option("--stack <stack>", StackName, "default")
  .on("--help", () => {
    console.log(docs("backup.md"));
  })
  .action(async () => {
    return;
  });
