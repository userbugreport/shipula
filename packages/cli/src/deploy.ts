import { Command } from "commander";
import { StackName } from "./constants";
import docs from "./docs";

export default new Command()
  .command("deploy [package]")
  .description("Deploys your pacakge to the cloud, creating and App and Stack.")
  .option("--stack <stack>", StackName, "default")
  .on("--help", () => {
    console.log(docs("deploy.md"));
  })
  .action(async () => {
    return;
  });
