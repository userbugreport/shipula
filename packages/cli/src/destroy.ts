import { Command } from "commander";
import { StackName } from "./constants";
import docs from "./docs";

export default new Command()
  .command("destroy [package]")
  .description("Completely remove a Stack from an App")
  .option("--stack <stack>", StackName, "default")
  .on("--help", () => {
    console.log(docs("destroy.md"));
  })
  .action(async () => {
    return;
  });
