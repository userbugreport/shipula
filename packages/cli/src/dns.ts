import { Command } from "commander";
import docs from "./docs";
import { buildEnvProps } from "./context";
import { DnsDomain } from "./components/DnsDomain";
import { display } from "./components/application";
import { EnvSet } from "./components/EnvSet";
import { Info } from "@shipula/context";

const domainCommand = new Command()
  .command("domain <domainName>")
  .action(async (domainName) => {
    await display({ domainName }, DnsDomain);
    return;
  });

const nameCommand = new Command()
  .command("name <hostName> <packageDirectory>] [stackName]")
  .action(
    async (hostName: string, packageDirectory: string, stackName: string) => {
      const props = await buildEnvProps(packageDirectory, stackName);
      props.setVariables = { SHIPULA_HOST_NAME: hostName };
      props.domainName = Info.domainName(hostName);
      // do the domain name verification
      await display(props, DnsDomain);
      // and then go ahead and update the env to set a host name
      await display(props, EnvSet);
      return;
    }
  );

export default new Command()
  .command("dns")
  .description("Manage custom domain names.")
  .on("--help", () => {
    console.log(docs("env.md"));
  })
  .addCommand(domainCommand)
  .addCommand(nameCommand);
