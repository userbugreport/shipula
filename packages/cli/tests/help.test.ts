import backup from "../src/backup";
import deploy from "../src/deploy";
import destroy from "../src/destroy";
import env from "../src/env";
import info from "../src/info";
import logs from "../src/logs";
import scale from "../src/scale";
import dns from "../src/dns";
import fs from "fs-extra";
import path from "path";

const commands = {
  backup,
  deploy,
  destroy,
  env,
  info,
  logs,
  scale,
  dns,
};

describe("CLI", () => {
  describe("sub commands", () => {
    describe.each(Object.keys(commands))("%s", (commandName) => {
      test("help strings", () => {
        const helpInfo = commands[commandName].helpInformation();
        expect(helpInfo).toMatchSnapshot();
        // and let's update the docs while we are here...
        fs.writeFileSync(
          path.join(
            __dirname,
            "..",
            "..",
            "docs",
            "docs",
            "commands",
            `${commandName}.mdx`
          ),
          `---
Title: ${commandName}
---

\`\`\`shell
${helpInfo}
\`\`\`
`
        );
      });
    });
  });
});
