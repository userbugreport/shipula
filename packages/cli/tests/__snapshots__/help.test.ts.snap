// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CLI sub commands backup help strings 1`] = `
"Usage:  backup [options] [package]

Creates a backup copy of your App/Stack file system, and provides a link where you can download it.

Options:
  --stack <stack>  Optional name to create another stack, or copy of your app -- like dev/test/stage (default: \\"default\\")
  -h, --help       display help for command
"
`;

exports[`CLI sub commands deploy help strings 1`] = `
"Usage:  deploy [options] [package]

Deploys your pacakge to the cloud, creating and App and Stack.

Options:
  --stack <stack>  Optional name to create another stack, or copy of your app -- like dev/test/stage (default: \\"default\\")
  -h, --help       display help for command
"
`;

exports[`CLI sub commands destroy help strings 1`] = `
"Usage:  destroy [options] [package]

Completely remove a Stack from an App

Options:
  --stack <stack>  Optional name to create another stack, or copy of your app -- like dev/test/stage (default: \\"default\\")
  -h, --help       display help for command
"
`;

exports[`CLI sub commands env help strings 1`] = `
"Usage:  env [options] [package]

Manage environment variables for an App/Stack.

Options:
  --stack <stack>  Optional name to create another stack, or copy of your app -- like dev/test/stage (default: \\"default\\")
  -h, --help       display help for command
"
`;

exports[`CLI sub commands info help strings 1`] = `
"Usage:  destroy [options] [package]

This will tell you all about your App.

Options:
  -h, --help  display help for command
"
`;

exports[`CLI sub commands logs help strings 1`] = `
"Usage:  logs [options] [package]

Stream real time logs or search cloud stored logs.

Options:
  --stack <stack>   Optional name to create another stack, or copy of your app -- like dev/test/stage (default: \\"default\\")
  --search <query>  Search for this content in log messages, returning matches
  -h, --help        display help for command
"
`;

exports[`CLI sub commands scale help strings 1`] = `
"Usage:  scale [options] [package]

Set scaling limits on your App / Stack.

Options:
  --stack <stack>  Optional name to create another stack, or copy of your app -- like dev/test/stage (default: \\"default\\")
  -h, --help       display help for command
"
`;
